src/Features/Core/Portable/CodeRefactorings/AddMissingImports/AddMissingImportsRefactoringProvider.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Composition;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.PasteTracking;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using static Microsoft.CodeAnalysis.CodeActions.CodeAction;
+
+namespace Microsoft.CodeAnalysis.AddMissingImports
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, LanguageNames.VisualBasic, Name = PredefinedCodeRefactoringProviderNames.AddMissingImports), Shared]
+    internal class AddMissingImportsRefactoringProvider : CodeRefactoringProvider
+    {
+        private readonly IPasteTrackingService _pasteTrackingService;
+
+        [ImportingConstructor]
+        public AddMissingImportsRefactoringProvider(IPasteTrackingService pasteTrackingService)
+        {
+            _pasteTrackingService = pasteTrackingService;
+        }
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+
+            if (!_pasteTrackingService.TryGetPasteTrackingInformation(document, out var trackingInformation)
+                || !context.Span.IntersectsWith(trackingInformation.TextSpan))
+            {
+                return;
+            }
+
+            var addMissingImportsService = document.GetLanguageService<IAddMissingImportsFeatureService>();
+
+            var isMissingImports = await addMissingImportsService.IsMissingImportsAsync(document, trackingInformation.TextSpan, context.CancellationToken).ConfigureAwait(false);
+            if (!isMissingImports)
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(new SolutionChangeAction(FeaturesResources.Add_missing_imports_for_pasted_code, async (cancellationToken) =>
+            {
+                var newProject = await addMissingImportsService.AddMissingImportsAsync(document, trackingInformation.TextSpan, context.CancellationToken).ConfigureAwait(false);
+                return newProject.Solution;
```

Comment:
given that this is returning a project, i assume this means this may end up adding *references* or *nuget packages*.  if so, it's likely the case that the user needs to more explicitly state that that's what tehy want.  We don't add-usings for references/nuget by ddefault.  We make a low-pri fix to offer that.

So maybe we have "Add missing imports yadda yadda" and "Add missing imports and references for yadda yadad".